diffprivlib/models/tree.py

```diff
@@ -0,0 +1,602 @@
+
+    References
+    ----------
+    [1] Sam Fletcher, Md Zahidul Islam. "Differentially Private Random Decision Forests using Smooth Sensitivity"
+    https://arxiv.org/abs/1606.03572
+
+    """
+
+    def __init__(self, n_estimators=10, epsilon=1.0, cat_feature_threshold=10,
+                 n_jobs=1, verbose=0, accountant=None, max_depth=15, random_state=None, feature_domains=None, **unused_args):
+        super().__init__(
+            base_estimator=DecisionTreeClassifier(),
+            n_estimators=n_estimators,
+            estimator_params=("cat_feature_threshold", "max_depth", "epsilon", "random_state"),
+            n_jobs=n_jobs,
+            random_state=random_state,
+            verbose=verbose)
+        self._n_estimators = n_estimators
+        self._n_jobs = n_jobs
+        self._epsilon = epsilon
+        self._cat_feature_threshold = cat_feature_threshold
+        self._random_state = random_state
+        self._estimators = []
+        self._verbose = verbose
+        self._max_depth = max_depth
+        self._accountant = BudgetAccountant.load_default(accountant)
+        self._n_features = None
+        self._classes = None
+        self._cat_features = None
+        self._feature_domains = feature_domains
+
+        if random_state is not None:
+            np.random.seed(random_state)
+
+        warn_unused_args(unused_args)
+
+    @property
+    def n_features(self):
+        return self._n_features
+
+    @property
+    def n_classes(self):
+        return self._n_classes
+
+    @property
+    def classes_(self):
+        return self._classes
+
+    @property
+    def cat_features(self):
+        return self._cat_features
+
+    @property
+    def max_depth(self):
+        return self._max_depth
+
+    @property
+    def estimators(self):
+        return self._estimators
+
+    @property
+    def feature_domains(self):
+        return self._feature_domains
+
+    def fit(self, X, y):
+        """Fit the model to the given training data.
+
+        Parameters:
+            X : array-like, shape (n_samples, n_features)
+                Training vector, where n_samples is the number of samples and n_features is the number of features.
+
+            y : array-like, shape (n_samples,)
+                Target vector relative to X.
+
+        Returns:
+            self: class object
+        """
+        if not isinstance(self._n_estimators, numbers.Integral) or self._n_estimators < 0:
+            raise ValueError(f'Number of estimators should be a positive integer; got {self._n_estimators}')
+
+        if not isinstance(self._cat_feature_threshold, numbers.Integral) or self._cat_feature_threshold < 0:
+            raise ValueError('Categorical feature threshold should be a positive integer;'
+                             f'got {self._cat_feature_threshold}')
+
+        self._accountant.check(self._epsilon, 0)
+
+        X, y = np.array(X), np.array(y)
+        check_X_y(X, y)
+
+        self._n_features = X.shape[1]
```

Comment:
Instead of `self._n_features`, use `self.n_features_in_`, as `n_features_` will raise an error in some versions of scikit-learn.